### PR TITLE
Add Discrete Exterior Calculus (DEC) operators and unit tests

### PR DESCRIPTION
**Pull request by mhajij**
This PR adds Discrete Exterior Calculus (DEC) support to TopoNetX, together with a unit test suite for it.

What was added

A user-facing ExteriorCalculusOperators class providing:

Coboundaries, Hodge stars, codifferentials, and DEC Hodge Laplacians

Geometry-free and geometry-aware metric backends

Metric and geometry backends for triangle meshes embedded in ℝ³

Unit tests covering:

Algebraic DEC identities

Geometry consistency and degeneracy handling

FEM mass, stiffness, and cotangent operators

Error handling and edge cases

Why DEC in TopoNetX

TopoNetX already provides rich combinatorial structures (simplicial, cell, and combinatorial complexes).
DEC is the natural calculus on these objects, enabling:

Structure-preserving discretization of differential operators

Intrinsic, coordinate-free PDEs on complexes and meshes

A principled bridge between topology, geometry, and numerical analysis

Direct support for physics-based simulation, FEM-style operators, and geometric learning

### CODE REVIEW

**Comment by mhajij** [comment]
@ffl096 please try to merge as soon as you can to avoid deterioration from main branch

File: toponetx/algorithms/exterior_calculus/exterior_calc.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+- topological coboundaries `d_k`,
+- diagonal Hodge stars `*_k`,
+- codifferentials `delta_k`,
+- DEC Hodge Laplacians `Delta_k`.
+
+For triangle meshes embedded in R^3, additional surface FEM operators on 0-forms
+are available through the chosen metric backend.
+"""
+
+from __future__ import annotations
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/exterior_calc.py]: I couldn't find a reason why you need to import future annotation support in this module?

File: toponetx/algorithms/exterior_calculus/metric.py [hunk 2]
```diff
@@ -0,0 +1,756 @@
...
+    return g0.astype(float), g1.astype(float), g2.astype(float), det_norm
+
+
+@dataclass(frozen=True)
+class MetricSpec:
+    """Describe how to build diagonal Hodge stars and anisotropic FEM tensors.
+
+    Attributes
+    ----------
+    preset : {"identity","diagonal","barycentric_lumped","circumcentric","voronoi","euclidean"}
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/metric.py]: Spaces between each element of the literal

File: toponetx/algorithms/exterior_calculus/metric.py [hunk 3]
```diff
@@ -0,0 +1,756 @@
...
+        Preset name controlling the backend.
+    diagonal_weights : dict[int, np.ndarray] | None
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/metric.py]: `, optional` instead of `| None` here (and for the other attributes below).

File: toponetx/algorithms/exterior_calculus/metric.py [hunk 4]
```diff
@@ -0,0 +1,756 @@
...
+    dual_measures: dict[int, np.ndarray] | None = None
+    tensors: np.ndarray | None = None
+    fn: MetricCallable | None = None
+    eps: float = 1e-12
+
+
+class HodgeStarBackend(Protocol):
+    """Protocol for Hodge star backends."""
+
+    def star(self, sc: Any, k: int, *, inverse: bool = False) -> csr_matrix:
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/metric.py]: `sc` is probably not of type `Any`

File: toponetx/algorithms/exterior_calculus/metric.py [hunk 5]
```diff
@@ -0,0 +1,756 @@
...
+            If True, return (*_k)^{-1}.
+
+        Returns
+        -------
+        csr_matrix
+            Sparse star matrix.
+        """
+
+
+def _as_1d_float(x: np.ndarray, *, name: str) -> np.ndarray:
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/metric.py]: It feels weird to require a `name` parameter just to be used in an error message. Whats the point? Context should be sufficiently given in the stacktrace?

**Review by ffl096 — CHANGES_REQUESTED** [review]
Just programming comments, I did not check the implementation for errors.

**Comment by mhajij** [comment]
> Just programming comments, I did not check the implementation for errors.

For this comment, I am adding next a tutorial that verifies the library against ground truth for some PDEs. This should be enough as it almost touches all important functions inside the libeary

File: test/algorithms/test_exterior_calc.py [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+- correct construction of DEC operators (coboundaries, Hodge stars, codifferentials),
+- valid shapes, basic algebraic identities, and symmetry where expected,
+- error handling for invalid inputs and unsupported configurations,
+- triangle-mesh metric presets route to the geometry backend and expose FEM helpers.
+
+These tests are intentionally lightweight and avoid asserting PDE convergence.
+They aim to validate correctness of API behavior and fundamental operator properties.
+"""
+
+from __future__ import annotations
```
Comment (ffl096) [inline test/algorithms/test_exterior_calc.py]: Why is the import necessary?

File: test/algorithms/test_exterior_calc.py [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+
+import numpy as np
+import pytest
+from scipy.sparse import csr_matrix, spmatrix
+
+import toponetx as tnx
```
Comment (ffl096) [inline test/algorithms/test_exterior_calc.py]: Please only use explicit imports in the package itself. The `import toponetx as tnx` import should only be used by end-users of the package.

File: test/algorithms/test_exterior_calc.py [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+        d0 = ops.d_matrix(0)
+        d1 = ops.d_matrix(1)
+
+        assert isinstance(d0, spmatrix)
+        assert isinstance(d1, spmatrix)
+
+        n0, n1, n2 = _counts(sc)
+        assert d0.shape == (n1, n0)  # C^0 -> C^1
+        assert d1.shape == (n2, n1)  # C^1 -> C^2
+        assert n2 == 1
```
Comment (ffl096) [inline test/algorithms/test_exterior_calc.py]: Please not only test that the shapes are correct, but the actual content of the matrices is.

File: test/algorithms/test_exterior_calc.py [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+        assert np.all(S0.diagonal() >= 0.0)
+        assert np.all(S1.diagonal() >= 0.0)
+        assert np.all(S2.diagonal() > 0.0)
+
+    def test_triangle_mesh_metric_requires_positions_raises(self):
+        """Raise when triangle-mesh metric is used without vertex positions."""
+        sc = build_single_triangle_sc_no_positions()
+
+        # Constructing the backend occurs during ops init; missing position labels should fail.
+        with pytest.raises((KeyError, ValueError, AttributeError)):
```
Comment (ffl096) [inline test/algorithms/test_exterior_calc.py]: I don't think you should expect all three error classes? Better to be specific.

File: test/algorithms/test_metric.py [hunk 10]
```diff
@@ -0,0 +1,820 @@
...
+            sc_obj : Any
+                Complex object.
+
+            Returns
+            -------
+            np.ndarray
+                SPD tensor of shape (3, 3).
+            """
+            _ = P
+            _ = sc_obj
```
Comment (ffl096) [inline test/algorithms/test_metric.py]: Why?

File: toponetx/algorithms/exterior_calculus/exterior_calc.py [hunk 11]
```diff
@@ -0,0 +1,891 @@
...
+        2. Normalizes the user-supplied ``metric`` into a
+           :class:`~toponetx.algorithms.exterior_calculus.metric.MetricSpec`
+           (or ``None`` to disable metric support).
+        3. If no explicit ``star_backend`` is provided, selects one automatically
+           based on the normalized metric preset.
+
+        Returns
+        -------
+        None
+            This method mutates ``metric`` and ``star_backend`` in place.
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/exterior_calc.py]: Remove this please.

File: toponetx/algorithms/exterior_calculus/exterior_calc.py [hunk 12]
```diff
@@ -0,0 +1,891 @@
...
+    @property
+    def dim(self) -> int:
+        """Return the maximum simplex dimension of the complex.
+
+        Returns
+        -------
+        int
+            Maximum simplex dimension.
+        """
+        return int(self.sc.dim)
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/exterior_calc.py]: Why the `int` cast? `self.sc.dim` should have the same type already, and if thats not the case in some circumstances, its a bug.

File: toponetx/algorithms/exterior_calculus/exterior_calc.py [hunk 13]
```diff
@@ -0,0 +1,891 @@
...
+        return self.sc.coincidence_matrix(k + 1, signed=signed, index=False)
+
+    def hodge_star(self, k: int, inverse: bool = False) -> csr_matrix:
+        r"""Return a discrete Hodge star matrix ``*_k``.
+
+        Parameters
+        ----------
+        k : int
+            Cochain degree.
+        inverse : bool
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/exterior_calc.py]:
```suggestion
        inverse : bool, default=False
```

File: toponetx/algorithms/exterior_calculus/exterior_calc.py [hunk 14]
```diff
@@ -0,0 +1,891 @@
...
+
+        Parameters
+        ----------
+        k : int
+            Cochain degree.
+        x : np.ndarray
+            Input k-cochain (1D array of length ``n_k``).
+        signed : bool
+            Whether to use signed coboundaries (orientation-aware).
+        solver : {"cg", "spsolve"}
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/exterior_calc.py]:
```suggestion
        solver : {"cg", "spsolve"}, default="cg"
```

File: toponetx/algorithms/exterior_calculus/exterior_calc.py [hunk 15]
```diff
@@ -0,0 +1,891 @@
...
+            Linear solver used for the normal equations.
+        tol : float
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/exterior_calc.py]:
```suggestion
        tol : float, default=1e-10
```

File: toponetx/algorithms/exterior_calculus/exterior_calc.py [hunk 16]
```diff
@@ -0,0 +1,891 @@
...
+                "riemannian_stiffness_0 requires a triangle-mesh metric preset."
+            )
+        return self.star_backend.riemannian_stiffness_0()
+
+    def cotan_laplacian_0(self, *, lumped: bool = True) -> csr_matrix:
+        r"""Return a cotangent Laplacian operator on 0-forms.
+
+        Parameters
+        ----------
+        lumped : bool
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/exterior_calc.py]:
```suggestion
        lumped : bool, default=True
```

File: toponetx/algorithms/exterior_calculus/metric.py [hunk 17]
```diff
@@ -0,0 +1,1040 @@
...
+
+from __future__ import annotations
+
+from collections import defaultdict
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any, Literal, Protocol
+
+import numpy as np
+from numpy.typing import NDArray
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/metric.py]: You most likely don't need `numpy.typing.NDArray` anymore. The standard `np.ndarray` is enough.

File: toponetx/algorithms/exterior_calculus/metric.py [hunk 18]
```diff
@@ -0,0 +1,1040 @@
...
+
+        Notes
+        -----
+        This constructs diagonal weights w_k = |dual_k| / |primal_k| using a barycentric dual.
+
+        Parameters
+        ----------
+        k : int
+            Cochain degree.
+        inverse : bool
```
Comment (ffl096) [inline toponetx/algorithms/exterior_calculus/metric.py]:
```suggestion
        inverse : bool, default=False
```

File: test/algorithms/test_exterior_calc.py [hunk 19]
```diff
@@ -0,0 +1,461 @@
...
+        L1 = ops.dec_hodge_laplacian_matrix(1)
+        L2 = ops.dec_hodge_laplacian_matrix(2)
+
+        assert isinstance(L0, spmatrix)
+        assert isinstance(L1, spmatrix)
+        assert isinstance(L2, spmatrix)
+
+        assert L0.shape == (n0, n0)
+        assert L1.shape == (n1, n1)
+        assert L2.shape == (n2, n2)
```
[the remark this answers is not in the document]
Comment (mhajij) [reply test/algorithms/test_exterior_calc.py]: This is for shape check, content is checked in other ways.

File: test/algorithms/test_metric.py (hunk 10 above)
Comment (mhajij) [reply test/algorithms/test_metric.py]: bad_fn is needed to test that invalid tensor shapes from user-provided metric functions are properly caught with a ValueError.
